### PR TITLE
Class-based transactions

### LINKED ISSUE #52: Class-based transactions
After working with dry-transaction for a while, and across a number of apps, I see a few issues in its design:

- The `Dry.Transaction` constructor creates objects with an ancestry that app authors have no control over, so there’s no way to add behaviour to all an app’s transactions (this is important in our apps at Icelab, because we use a little `#to_queue/#call_from_queue` protocol for running operations in background jobs.
- Also because of this, there’s no place to put transaction-specific logic for adjusting values in-between a transaction’s steps, which can be useful for tying together steps with slightly differing input/output expectations.
- The constructor returning anonymous objects means we need some sort of harness for building and capturing transactions (i.e. registering in our container) in each of our apps.
- Having to build and register transactions like this means that their definition ends up sitting far apart from other parts of the codebase that they relate to, and their registration ends up complicating the app’s boot process.

I think we can solve all of these by moving to **class-based transactions**.

After thinking about it for a while, our transactions are really just a specialised case for auto-injection: a transaction collects a series of operation objects (mapped to container identifiers), and calls them in sequence.

So instead of building a transaction like this:

```ruby
my_trans = Dry.Transaction(MyContainer) do
  step :one, "operations.one"
  step :two, "operations.two"
end
```

We could build one like this:

```ruby
class MyTransaction
  # Similar mixin to Dry::AutoInject
  include Dry::Transaction(MyContainer)
  
  step :one, "operations.one"
  step :two, "operations.two"
end

my_trans = MyTransaction.new
```

This would allow us to inject replacement dependencies to make testing easier:

```ruby
# Provide an explicit operation for step one
MyTransaction.new(one: different_operation)
```

For larger apps, this’d let us provide a per-container transaction mixin for easy reuse, too:

```ruby
# In container setup
module Main
  Transaction = Dry::Transaction(Main::Container)
end

# Elsewhere
module Main
  class AnotherTransaction
    include Main::Transaction
    
    step :one
    # …
  end
end
```

Or a base class:

```ruby
module Main
  class Transaction
    # mix Dry::Transaction module into subclasses

    # Can provide custom call logic if needed
    # def call
    # end
    
    # And shared API to transaction objects
    def to_queue(input)
      [input.id]
    end
    
    def call_from_queue(input_id)
      input = find_from_id(input_id)
      call(input)
    end
  end
end
```

Another benefit from class-based transactions is that the application author can add extra, local, steps if they needed to:

```ruby
class MyTransaction
  include Main::Transaction
  
  step :one
  step :prepare_two
  step :two
  
  # Can mix local methods with injected step operations
  def prepare_two(input)
    # do something with output of one
  end
end
```

Or even wrap the existing steps:

```ruby
class MyTransaction
  include Main::Transaction
  
  step :one
  step :two
  
  def one(input)
    changed_input = do_something_with(input)
    super(changed_input)
  end
end
```

And all the other stuff that we get with Ruby’s standard class behaviour.

I think this approach would make dry-transaction easier and more flexible to work with, would make it fit more naturally into larger apps, and handle a variety of different real-world scenarios, many of which we couldn’t even predict right now.

What do you think? (/cc @solnic, @flash-gordon, @AMHOL) Any objections to me working on this approach?

### PR DESCRIPTION
**Pull request by GustavoCaso**
Implement class-based transactions based on design outlined in #52. 

Resolves #52.

<hr>

_Original PR notice:_

This implements this #52 

This still experimental, but before adding more code and more logic to the library I wanted to share @timriley the simplest example so we can discuss how to continue working on this.

All feedback is gratefully received 😄 

Thanks

### CODE REVIEW

**Comment by GustavoCaso** [comment]
@timriley so far with this implementation we can replacement dependencies to make testing easier:

 ```ruby
# Provide an explicit operation for step one
MyTransaction.new(one: different_operation)
```

But I think this approach is not going to help because one key feature would be been able to wrap existing steps. I was thinking maybe we could instead of creating the step within the macro to store them in some class level variables like `step_info` that store step type `map, raw, tee` and the name of the step, with that in place at initialization time we could dynamically generate the instance method with the step names, that way the user can wrap them. 

Also since we want that the user can provide extra locals step, I was thinking of modifying the `Transaction` class to instead of getting `steps` and `matcher` it can get the instance of the class so we can create the steps checking for them in the container or in the class.

What do you think?

**Comment by timriley** [comment]
Hey @GustavoCaso, thanks for pushing this forward! I've spent some time this morning sketching out in code the kind of direction we probably wanna take this, but I feel I haven't got to a good spot either. You can see I've pushed up a branch here (https://github.com/dry-rb/dry-transaction/compare/class_base_transaction...class_based_transactions_timriley) - feel free to take a look and let me know if you have any thoughts so far. I feel I need to experiment just a tad more before I can summarise things nearly in words for you. Roughly, though:

- I've made this work in a way where we don't _require_ container objects as operations, since users can just provide local methods for some/all of their steps, if they want
- I'm not happy with where the container object resolution is, though. I feel like including (or prepending?) an `#initialize` might be good, which is the only place we try and resolve stuff, and then we can pass them through to the standard initialise for storing those objects in local vars

Sorry I can't get back to you with anything better so far, let's just keep the conversation going.

**Comment by GustavoCaso** [comment]
@timriley I added the method for passing extra step arguments, and update all integration test to use new class base transaction.

Also, remove all not used code to make things easier while reviewing this PR.

When we are to place we are happy to push this change, I can start with the documentation both in the project and in the `dry-rb.org` website.

File: lib/dry/transaction/step.rb [hunk 1]
```diff
@@ -29,8 +29,7 @@ def initialize(step_adapter, step_name, operation_name, operation, options, call
         @call_args = call_args
       end
 
-      # TODO: rename to curry
-      def apply(operation: UNDEFINED, call_args: UNDEFINED)
+      def curry(operation: UNDEFINED, call_args: UNDEFINED)
```
Comment (timriley) [inline lib/dry/transaction/step.rb]: I think my TODO here was actually out of date. I don't think curry as a name makes sense if we're passing more than just call args. Would you mind renaming again to `#with`? That's more consistent with similar methods elsewhere in dry-rb land.

File: lib/dry/transaction/step.rb [hunk 2]
```diff
@@ -27,12 +29,24 @@ def initialize(step_adapter, step_name, operation_name, operation, options, call
         @call_args = call_args
       end
 
-      def with_call_args(*call_args)
-        self.class.new(step_adapter, step_name, operation_name, operation, options, call_args, &block)
+      def curry(operation: UNDEFINED, call_args: UNDEFINED)
```
Comment (timriley) [inline lib/dry/transaction/step.rb]: I think my TODO about renaming this to `#curry` was actually out of date. I don't think curry as a name makes sense if we're passing more than just call args. Would you mind renaming again to `#with`? That's more consistent with similar methods elsewhere in dry-rb land.

**Comment by timriley** [comment]
This is looking pretty good, now! Here's what I think we have left:

- Merge all the stuff from `spec/integration/class_based_spec.rb` into `spec/integration/transaction_spec.rb` (since _every_ transaction is class-based now, I don't see the need to have a special spec file for this)
- Consider if we can bring back `#prepend`/`#append`/`#insert`/`#remove`. I have a feeling that some people will be using these, and if there's a way we can still support them, it'd be nice.

Once we've done all this, I might ping a few more people for review since this is a big change to how dry-transaction works.

I'll take a look at the TravisCI build stuff in a day or too, btw.

**Comment by timriley** [comment]
@GustavoCaso Made some more changes here. I decided that the append/prepend/insert stuff was too fraught with potential issues for us to keep it (especially when we're potentially mixing different transaction classes with their own different configurations/internal state/other private helper methods, etc.).

**Comment by timriley** [comment]
@GustavoCaso We did it! 🎉 

Merging this now and will hopefully release tomorrow along with some updated docs over in https://github.com/dry-rb/dry-rb.org/pull/180.

**Pull request merged**